File: src/contexts/core/kanji/domain/events/kanji-created-event.ts
import { DomainEvent } from "../../../../shared/events/domain/domain-event";
import type { DomainEventId } from "../../../../shared/events/domain/domain-event-id";
import type { DomainEventOccurredOn } from "../../../../shared/events/domain/domain-event-occurred-on";
import type { KanjiIdeogram } from "../valueobjects/kanji-ideogram";
import type { KanjiKunyomiReading } from "../valueobjects/kanji-kunyomi-reading";
import type { KanjiMeaning } from "../valueobjects/kanji-meaning";
import type { KanjiOnyomiReading } from "../valueobjects/kanji-onyomi-reading";
import type { KanjiRadical } from "../valueobjects/kanji-radical";
import type { KanjiStrokesNumber } from "../valueobjects/kanji-strokes-number";

class KanjiCreatedEvent extends DomainEvent {
    constructor(
        public readonly id: DomainEventId,
        public readonly occurredOn: DomainEventOccurredOn,
        public readonly ideogram: KanjiIdeogram,
        public readonly onyomiReadings: KanjiOnyomiReading[],
        public readonly kunyomiReadings: KanjiKunyomiReading[],
        public readonly meanings: KanjiMeaning[],
        public readonly strokes: KanjiStrokesNumber,
        public readonly radicals: KanjiRadical[],
    ) {
        super();
    }

    get type(): string {
        return 'tanukanji.core.core.kanji.created:1.0';
    }
}

export { KanjiCreatedEvent };

File: libs/kernel/src/kanjis/domain/events/kanji_created_event.rs
use std::sync::Arc;

use events::domain::domain_event::DomainEvent;

use crate::kanjis::domain::entities::{kanji_ideogram::KanjiIdeogram, kanji_kunyomi_reading::KanjiKunyomiReading, kanji_meaning::KanjiMeaning, kanji_nanori_reading::KanjiNanoriReading, kanji_onyomi_reading::KanjiOnyomiReading, kanji_radical::KanjiRadical, kanji_strokes_number::KanjiStrokesNumber};


pub struct KanjiCreatedEvent {
    pub id: String,
    pub ideogram: KanjiIdeogram,
    pub kunyomi_readings: Vec<KanjiKunyomiReading>,
    pub onyomi_readings: Vec<KanjiOnyomiReading>,
    pub nanori_readings: Vec<KanjiNanoriReading>,
    pub meanings: Vec<KanjiMeaning>,
    pub strokes: KanjiStrokesNumber,
    pub radicals: Vec<KanjiRadical>,
    pub occurred_on: String,
}

impl KanjiCreatedEvent {
    pub fn new(
        ideogram: KanjiIdeogram, 
        kunyomi_readings: Vec<KanjiKunyomiReading>, 
        onyomi_readings: Vec<KanjiOnyomiReading>, 
        nanori_readings: Vec<KanjiNanoriReading>, 
        meanings: Vec<KanjiMeaning>, 
        strokes: KanjiStrokesNumber, 
        radicals: Vec<KanjiRadical>,
    ) -> Self {
        let id = uuid::Uuid::new_v4().to_string();
        let occurred_on = chrono::Utc::now().to_rfc3339();
        Self { id, ideogram, kunyomi_readings, onyomi_readings, nanori_readings, meanings, strokes, radicals, occurred_on }
    }

    pub fn new_shared(
        ideogram: KanjiIdeogram, 
        kunyomi_readings: Vec<KanjiKunyomiReading>, 
        onyomi_readings: Vec<KanjiOnyomiReading>, 
        nanori_readings: Vec<KanjiNanoriReading>,
        meanings: Vec<KanjiMeaning>, 
        strokes: KanjiStrokesNumber, 
        radicals: Vec<KanjiRadical>
    )  -> Arc<Self> {
        std::sync::Arc::new(Self::new(ideogram, kunyomi_readings, onyomi_readings, nanori_readings, meanings, strokes, radicals))
    }
}

impl DomainEvent for KanjiCreatedEvent {
    fn event_type(&self) -> String {
        "tanukanji.kernel.kanji.created@1.0.0".to_string()
    }
}